docs: profiling howto [DET-5209] (#2384)
determined-ai/determined: docs/how-to/profiling.txt

.. _howto-profiling:

#################################
 How To: Profile Your Experiment
#################################

> When you want to improve the training speed of your model, the first
> thing you need to do is understand why training is slow. To do this, you
> need to profile your training code. There are many different layers you
> can look at, from raw throughput all the way to GPU kernals.

Reviewer: ```suggestion
When you want to debug the training speed of a model, you need
tools to inspect and measure internal states and runtime statistics of
the model during execution. There are many different layers you
can look at, from raw throughput all the way to GPU kernels.
```
Idk this is a hard paragraph to write.. i want to take a course on technical writing. I feel like there is a better way to motivate the reader but i can't quite find the words

Author: I rewrote it a bit. It could probably still use some improvements.

Determined provides two tools out-of-the-box for profiling your training

> - System Metrics and Timings.

Author: Fixed

****************
 System Metrics
****************

System Metrics are statistics around hardware usage, such as GPU
utilization and network throughput. These metrics are useful for seeing
whether your training is using the hardware effectively. When the System
Metrics reported for an experiment are below what you expect from your
hardware, that is a sign that the software may be able to be optimized
to make better use of the hardware resources.

Specifically, Determined tracks:

-  GPU utilization
-  GPU memory usage
-  Network throughput (sent)
-  Network throughput (received)
-  Disk IOPS
-  Disk throughput (read)
-  Disk throughput (write)
-  Host memory sage
-  Average CPU utilization across cores

For distributed training, these metrics are collected for every agent.
The data is broken down per agent. GPU metrics can be further broken
down by GPU.

> Note: System Metrics record agent-level metrics, so when there are

Reviewer: I'd use a
```
.. note::
```
block

multiple experiments on the agent, it is difficult to analyze. We
suggest that profiling is done when the experiment is the only thing
running on the agent.

> *******************************

Reviewer: I could take or leave this, or add it to a follow up if you wanted (seems it this could be a topic guide on profiling in determined and the other (this) could be a how to?).

Author: I removed it. It can be an add on-later

 Example: Gradient Aggregation
*******************************

TODO: Walk through how to use System Metrics to look at Gradient
Aggregation

*********
 Timings

> *********

Reviewer: when @azhou-determined is done with the profiling tools work, maybe we could add a section here for "Native Tooling"

> The other class of profiling that Determined handles is Timings. This
> tracks how long training events take, such as retrieving data from the
> data loader, moving data host-to-device, running the forward/backwards
> pass, or executing callbacks.

Reviewer: ```suggestion
The other type of profiling metric that Determined tracks are Timings. These
track how long training events take, such as retrieving data from the
data loader, moving data host-to-device, running the forward/backwards
pass, or executing callbacks.
```

Author: I improved this, although it's slightly different than this suggestion

This provides a fairly high-level picture of where you might want to
spend your time further investigating the code.

Specifically, Determined tracks:

> -  ``dataloader_next`` (amount of time to retrieve the next item from

Reviewer: ```suggestion
-  ``dataloader_next`` (retrieving the next item from
```
and similar maybe?

Author: fixed

   the dataloader)
-  ``to_device`` (amount of time transfering input from host to device)
-  ``train_batch`` (amount of time user-defined ``train_batch`` function
   takes to execute\*)
-  ``step_lr_schedulers`` (amount time taken to update the LR schedules)
-  ``from_device`` (amount of time transfering output from device to
   host)
-  ``reduce_metrics`` (amount of time taken to calculate global metrics
   in distributed training)

> \* ``train_batch`` is typically the forward pass and the backwards pass.

Reviewer: typo?

Author: How so?

> Timings are currently only supported in PyTorch.

Reviewer: I may move this to the intro paragraph, or a note below it.

Author: Fixed

.. _howto-profiling:

#################################

>  How To: Profile An Experiment

Author: I think this page naming convention is useful, particularly for search, but it is inconsistent with the rest of the pages so maybe we should replace the title.

Reviewer: you mean the prefix of "how to"?

Author: Yeah

-  :ref:`System Metrics <howto-profiling-system-metrics>`: measurements
   of hardware usage
-  :ref:`Timings <howto-profiling-timings>`: durations of actions taken
   during training, such as dataloading

System Metrics are useful to see if the software is taking full
advantage of the available hardware, particularly around GPU usage,
dataloading, and network communication during distributed training.
Timings are useful for identifying the section of code to focus on for
optimizations. Most commonly, Timings help answers the question of
whether the dataloader is the main bottleneck in training.

.. _howto-profiling-system-metrics:

****************
 System Metrics
****************

System Metrics are statistics around hardware usage, such as GPU
utilization and network throughput. These metrics are useful for seeing
whether training is using the hardware effectively. When the System
Metrics reported for an experiment are below what is expected from the
hardware, that is a sign that the software may be able to be optimized
to make better use of the hardware resources.

Specifically, Determined tracks:

-  GPU utilization

> -  GPU memory usage

Reviewer: should we say free mem here and below? or like `GPU memory usage (as the amount of free GPU memory)`

Author: I updated it to be more accurate - "GPU free memory" and "Host available memory"
